CA-340484 bump xapi version in DB after upgrade/update

lippirk opened the pull request:
We got round to bumping the hardcoded API version in b81af6e, but
the API version is also stored in the DB, which was not changed.
This was causing problems with the firstboot create-guest-template
service since ultimately it gets a host's API version from the DB,
and then uses that version to perform a template import, leading to
an IMPORT_INCOMPATIBLE_VERSION error.

lippirk commented:
testing in progress

edwintorok commented on ocaml/xapi/xapi_db_upgrade.ml:
```diff
@@ -612,6 +612,17 @@ let remove_legacy_ssl_support = {
+  fn=fun ~__context ->
+    Db.Host.get_all ~__context |>
+    List.iter (fun self ->
+      Db.Host.set_API_version_major ~__context ~self ~value:(Int64.of_int Constants.version_major);
```
What happens during an RPU when the master got upgraded, but the slave hasn't yet? This looks more like a pool version than a host version now.
If this causes other problems could we set the API version on toolstack startup just for the current host? That'd give a better picture...

edwintorok replied:
Dbsync_slave refresh_localhost_info already has some code to do this refresh (and apparently runs on master too).
Is the problem that the import code runs too soon, before that version is updated?

lippirk commented:
as @psafont pointed out, i was updating the wrong version, so I've fixed that now

i moved the version update to `refresh_localhost_info`. `create-guest-templates.service` ought run after this, because it waits for `xapi-init-complete.service`

lindig commented on ocaml/xapi/create_misc.mli:
```diff
@@ -44,3 +44,4 @@ val create_host_cpu : __context:Context.t -> host_info -> unit
 val create_pool_cpuinfo : __context:Context.t -> unit
 val create_chipset_info : __context:Context.t -> host_info -> unit
 val create_updates_requiring_reboot_info : __context:Context.t -> host:[`host] Ref.t -> unit
+val get_xapi_verstring : unit -> string
```
`get_xapi_verstring` - the fact that this is a string is covered by the type, so I would prefer `get_xapi_version`. This will lead to clearer code below because "version" will be used more consistently.

edwintorok commented on ocaml/xapi/dbsync_slave.ml:
```diff
@@ -84,6 +84,18 @@ let refresh_localhost_info ~__context info =
+    let current_xapi_version = Create_misc.get_xapi_version () in
+    if old_xapi_version <> current_xapi_version then
+      let value = ("xapi", current_xapi_version) :: List.remove_assoc "xapi" old_software_version in
+      Db.Host.set_software_version ~__context ~self:host ~value
```
Hmm, didn't Create_misc.create_software_version above already call this?

lippirk commented:
this is the wrong fix, closing
